src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -597,6 +649,666 @@ HRESULT STDMETHODCALLTYPE CorProfiler::JITCompilationStarted(
+      asmRefInfo.cbPublicKeyOrToken = public_key_size;
+      asmRefInfo.szName = assemblyReference.name.c_str();
+      asmRefInfo.pMetaData = &assembly_metadata;
+      asmRefInfo.pbHashValue = nullptr;
+      asmRefInfo.cbHashValue = 0;
+      asmRefInfo.dwAssemblyRefFlags = 0;
+
+      auto hr = pAsmRefProvider->AddAssemblyReference(&asmRefInfo);
+
+      // This currently returns 0x8007000b (ERROR_BAD_FORMAT: An attempt was made to load a program with an incorrect format.)
+      if (FAILED(hr)) {
+        Warn("GetAssemblyReferences failed for call from ", wszAssemblyPath);
+        return S_OK;
+      }
+
+      Debug("GetAssemblyReferences extending assembly closure for ",
+           assembly_name, " to include", asmRefInfo.szName);
+      return S_OK;
+    }
+  }
+
+  return S_OK;
+}
+
 bool CorProfiler::IsAttached() const { return is_attached_; }
 
+//
+// Startup methods
+//
+HRESULT CorProfiler::RunILStartupHook(const ComPtr<IMetaDataEmit2>& metadata_emit,
+                                      const ModuleID module_id,
+                                      const mdToken function_token) {
+  mdMethodDef ret_method_token;
+  auto hr = GenerateVoidILStartupMethod(module_id, &ret_method_token);
+  if (FAILED(hr)) {
+    Warn("RunILStartupHook: Call to GenerateVoidILStartupMethod failed for ", module_id);
+    return S_OK;
+  }
+
+  ILRewriter rewriter(this->info_, nullptr, module_id, function_token);
+  hr = rewriter.Import();
+  RETURN_OK_IF_FAILED(hr);
+
+  ILRewriterWrapper rewriter_wrapper(&rewriter);
+
+  // Get first instruction and set the rewriter to that location
+  ILInstr* pInstr = rewriter.GetILList()->m_pNext;
+  rewriter_wrapper.SetILPosition(pInstr);
+  rewriter_wrapper.CallMember(ret_method_token, false);
+  hr = rewriter.Export();
+  RETURN_OK_IF_FAILED(hr);
+
+  return S_OK;
+}
+
+HRESULT CorProfiler::GenerateVoidILStartupMethod(const ModuleID module_id,
+                                                 mdMethodDef* ret_method_token) {
+  ComPtr<IUnknown> metadata_interfaces;
+  auto hr = this->info_->GetModuleMetaData(module_id, ofRead | ofWrite,
+                                           IID_IMetaDataImport2,
+                                           metadata_interfaces.GetAddressOf());
+  if (FAILED(hr)) {
+    Warn("GenerateVoidILStartupMethod: failed to get metadata interface for ", module_id);
+    return hr;
+  }
+
+  const auto metadata_import =
+      metadata_interfaces.As<IMetaDataImport2>(IID_IMetaDataImport);
+  const auto metadata_emit =
+      metadata_interfaces.As<IMetaDataEmit2>(IID_IMetaDataEmit);
+  const auto assembly_import = metadata_interfaces.As<IMetaDataAssemblyImport>(
+      IID_IMetaDataAssemblyImport);
+  const auto assembly_emit =
+      metadata_interfaces.As<IMetaDataAssemblyEmit>(IID_IMetaDataAssemblyEmit);
+
+  // TODO fix this for .NET Core
+  // Define an AssemblyRef to mscorlib, needed to create TypeRefs later
+  mdModuleRef mscorlib_ref;
+  ASSEMBLYMETADATA metadata;
```

Comment:
You can initializes values to defaults by using one of the following instead of `ZeroMemory()`:
```cpp
ASSEMBLYMETADATA metadata();
ASSEMBLYMETADATA metadata = {};
ASSEMBLYMETADATA metadata{};    // (since C++11)
```

https://en.cppreference.com/w/cpp/language/zero_initialization

```suggestion
  ASSEMBLYMETADATA metadata{};
```